src/vos/ilog.c

```diff
@@ -597,96 +495,73 @@ ilog_destroy(struct umem_instance *umm,
 static int
 ilog_root_migrate(struct ilog_context *lctx, const struct ilog_id *id_in)
 {
-	union prec		 prec;
-	daos_epoch_t		 epoch;
-	struct ilog_root	*root;
 	struct ilog_root	 tmp = {0};
-	d_iov_t			 key_iov;
-	d_iov_t			 val_iov;
 	umem_off_t		 tree_root;
-	daos_handle_t		 toh = DAOS_HDL_INVAL;
-	struct umem_attr	 uma;
-	struct ilog_id		 id = *id_in;
+	struct ilog_root	*root;
+	struct ilog_array	*array;
 	int			 rc = 0;
+	int			 idx;
 
 	root = lctx->ic_root;
 
 	rc = ilog_tx_begin(lctx);
 	if (rc != 0) {
 		D_ERROR("Failed to start PMDK transaction: rc = %s\n",
 			d_errstr(rc));
-		goto done;
+		return rc;
 	}
 
-	umem_attr_get(&lctx->ic_umm, &uma);
-	rc = dbtree_create(VOS_BTR_ILOG, BTR_FEAT_UINT_KEY, ILOG_TREE_ORDER,
-			   &uma, &tree_root, &toh);
-	if (rc != 0) {
-		D_ERROR("Failed to create an incarnation log tree: rc = %s\n",
-			d_errstr(rc));
-		goto done;
-	}
+	tree_root = umem_alloc(&lctx->ic_umm, sizeof(*array) + sizeof(array->ia_id[0]) * 2);
```

Comment:
> My main concern is allocating contiguous pmem buffer, but I guess it might not be a big deal because our DTX batch commit threshold is just a few of hundreds?
> However, if the system is very busy and batch commit is lagging behind, then batch commit and aggregation might lag behind I/Os, then we could end up with large amount of log entries?

We still have the cos code that prevents large ilog in practice because it piggybacks commits with conflicting updates.   So I don't think it ever gets very long.   @Nasf-Fan can confirm or correct.

Reply:
> it's allocating 96 bytes at here, how about make it 128 because I assume it consumes 128 bytes anyway (PMDK allocator).

Good point.  Maybe I can start with a larger array to align it and reduce allocations

Reply:
I checked this.  It's actually 40 bytes.  struct ilog_id is 16 bytes and ilog_array is 8 bytes.   If I add some padding to ilog_array, I can always increase by 128 bytes with 7 initial entries and 8 additional entries per resize.  Or I can continue to double each time.